src/ucp/core/ucp_ep.inl

```diff
@@ -156,42 +156,49 @@ static UCS_F_ALWAYS_INLINE ucp_ep_flush_state_t* ucp_ep_flush_state(ucp_ep_h ep)
     return &ucp_ep_ext_gen(ep)->flush_state;
 }
 
-static UCS_F_ALWAYS_INLINE uintptr_t ucp_ep_dest_ep_ptr(ucp_ep_h ep)
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t ucp_ep_rkey(ucp_ep_h ep)
 {
 #if UCS_ENABLE_ASSERT
-    if (!(ep->flags & UCP_EP_FLAG_DEST_EP)) {
-        return 0; /* Let remote side assert if it gets NULL pointer */
+    if (!(ep->flags & UCP_EP_FLAG_RKEY)) {
+        /* Let remote side assert if it gets invalid key */
+        return UCP_WORKER_PTR_KEY_INVALID;
     }
 #endif
-    return ucp_ep_ext_gen(ep)->dest_ep_ptr;
+    return ucp_ep_ext_gen(ep)->keys->remote;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_ptr_map_key_t ucp_ep_lkey(ucp_ep_h ep)
+{
+    ucs_assert(ucp_ep_ext_gen(ep)->keys->local != UCP_WORKER_PTR_KEY_INVALID);
+    return ucp_ep_ext_gen(ep)->keys->local;
 }
 
 /*
  * Make sure we have a valid dest_ep_ptr value, so protocols which require a
  * reply from remote side could be used.
  */
 static UCS_F_ALWAYS_INLINE ucs_status_t
-ucp_ep_resolve_dest_ep_ptr(ucp_ep_h ep, ucp_lane_index_t lane)
+ucp_ep_resolve_rkey(ucp_ep_h ep, ucp_lane_index_t lane)
 {
-    if (ep->flags & UCP_EP_FLAG_DEST_EP) {
+    if (ep->flags & UCP_EP_FLAG_RKEY) {
         return UCS_OK;
     }
 
     return ucp_wireup_connect_remote(ep, lane);
 }
 
-static inline void ucp_ep_update_dest_ep_ptr(ucp_ep_h ep, uintptr_t ep_ptr)
+static inline void ucp_ep_update_rkey(ucp_ep_h ep, ucs_ptr_map_key_t rkey)
 {
-    if (ep->flags & UCP_EP_FLAG_DEST_EP) {
-        ucs_assertv(ep_ptr == ucp_ep_ext_gen(ep)->dest_ep_ptr,
-                    "ep=%p ep_ptr=0x%lx ep->dest_ep_ptr=0x%lx",
-                    ep, ep_ptr, ucp_ep_ext_gen(ep)->dest_ep_ptr);
+    if (ep->flags & UCP_EP_FLAG_RKEY) {
+        ucs_assertv(rkey == ucp_ep_ext_gen(ep)->keys->remote,
+                    "ep=%p rkey=0x%lx ep->rkey=0x%lx",
+                    ep, rkey, ucp_ep_ext_gen(ep)->keys->remote);
```

Comment:
maybe return here?

Reply:
I did not change this logic, I guess there was assumption that this branch has to be compiled out in release.